Linux and mac problem with --watch option with a tsconfig.json « "include": ["app/**/*.ts"] »

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:** 2.8.0-dev.20180127 or 2.6.2

**OS:** Linux x86_64

**Node Version:** 9.4.0

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 
* "is:issue is:open watch"

**Steps to reproduce:**

1. `mkdir ~/essaiTS && cd $_`

2 . ~/essaiTS/tsconfig.json
```json
{
    "include": [
        "app/**/*.ts"
    ]
}
```
3. ~/essaiTS/app/file.ts
```typescript
console.log('aaa');
```
4. ` node node_modules/typescript/bin/tsc --watch --listEmittedFiles`
    * We could see a new file ~/essaiTS/app/file.js with:
```javascript
console.log('aaa');
```
5.  ~/essaiTS/app/file.ts
```typescript
console.log('aaa');
console.log('bbb');
```
**Expected behavior:**
At  ~/essaiTS/app/file.js. 
we must have:
```javascript
console.log('aaa');
console.log('bbb');
```
**Actual behavior:**
As in step 4., we have instead only:
```javascript
console.log('aaa');
```
With TypeScript 6.2 we could see in terminal : 
```
3:14:46 PM - File change detected. Starting incremental compilation...
3:14:47 PM - Compilation complete. Watching for file changes.
```
~~With TypeScript 2.8.0-dev.20180127 there is no messages.~~ (error on my part, with TypeScript 2.8.* console is just cleared: see my following comment.)
**Notes:**
* No problems with compilation option `--watch` if essaiTS is in folder /tmp. 
    To test you could use command line: `cp -R ~/essaiTS /tmp && cd /tmp/essaiTS && node node_modules/typescript/bin/tsc --watch --listEmittedFiles`, then edit file /tmp/essaiTS/app/file.ts.
* There is no problems with option `--watch` for files in  ~/essaiTS/app/folder/.

**Related Issues:**
Maybe: https://github.com/Microsoft/TypeScript/issues/20739

Yes, I've seen it's a big merge.

I've explored more thanks git bisect but it's hard because there are lot of changes, lot of break, and there is a lot of changes.

```
89c61e797c794515a06a3719ca95e37d97783d6e is the first bad commit
commit 89c61e797c794515a06a3719ca95e37d97783d6e
Author: Sheetal Nandi <shkamat@microsoft.com>
Date:   Thu Aug 3 00:27:46 2017 -0700

    Modify the api in builder so that it tracks changed files
````

At 89c61e797c794515a06a3719ca95e37d97783d6e, you could see an error message when you use incremental compilation (see message below). It was introduced by   c814d8e8 and was resolved by 59d07dc.  

`$ node ~/essaiTS/lib/tsc.js --watch`, then edit and save the file ~/essaiTS/app/file.ts.
```

9:31:02 PM - Compilation complete. Watching for file changes.


error TS18003: No inputs were found in config file '/home/.........../essaiTS/tsconfig.json'. S
pecified 'include' paths were '["app/**/*.ts"]' and 'exclude' paths were '[]'.
error TS18003: No inputs were found in config file '/home/............./essaiTS/tsconfig.json'. S
pecified 'include' paths were '["app/**/*.ts"]' and 'exclude' paths were '[]'.
9:31:09 PM - File change detected. Starting incremental compilation...


9:31:09 PM - Compilation complete. Watching for file changes.


error TS18003: No inputs were found in config file '/home/.............../essaiTS/tsconfig.json'. S
pecified 'include' paths were '["app/**/*.ts"]' and 'exclude' paths were '[]'.
error TS18003: No inputs were found in config file '/home/.................../essaiTS/tsconfig.json'. S
pecified 'include' paths were '["app/**/*.ts"]' and 'exclude' paths were '[]'.
9:31:24 PM - File change detected. Starting incremental compilation...


9:31:24 PM - Compilation complete. Watching for file changes.
```

* (Note: from e068475 to 69e5abd5b7(parent of c814d8e8) I've found some crashes. The watching compilation doesn't start with different errors).

I hope now it's ok !

--- a/src/compiler/watch.ts
+++ b/src/compiler/watch.ts
@@ -420,6 +420,9 @@ namespace ts {
         }
     }
 
+    const intialVersion = 1;
+    const intialVersionString = "1";
+
     /**
      * Creates the watch from the host for root files and compiler options
      */
@@ -429,11 +432,17 @@ namespace ts {
      */
     export function createWatchProgram<T extends BuilderProgram>(host: WatchCompilerHostOfConfigFile<T>): WatchOfConfigFile<T>;
     export function createWatchProgram<T extends BuilderProgram>(host: WatchCompilerHostOfFilesAndCompilerOptions<T> & WatchCompilerHostOfConfigFile<T>): WatchOfFilesAndCompilerOptions<T> | WatchOfConfigFile<T> {
-        interface HostFileInfo {
+        interface FilePresentOnHost {
             version: number;
             sourceFile: SourceFile;
             fileWatcher: FileWatcher;
         }
+        type FileMissingOnHost = string;
+        interface FilePresenceUnknownOnHost {
+            version: number;
+        }
+        type FileMayBePresentOnHost = FilePresentOnHost | FilePresenceUnknownOnHost;
+        type HostFileInfo = FilePresentOnHost | FileMissingOnHost | FilePresenceUnknownOnHost;
 
         let builderProgram: T;
         let reloadLevel: ConfigFileProgramReloadLevel;                      // level to indicate if the program needs to be reloaded from config file/just filenames etc
@@ -441,7 +450,7 @@ namespace ts {
         let watchedWildcardDirectories: Map<WildcardDirectoryWatcher>;      // map of watchers for the wild card directories in the config file
         let timerToUpdateProgram: any;                                      // timer callback to recompile the program
 
-        const sourceFilesCache = createMap<HostFileInfo | string>();        // Cache that stores the source file and version info
+        const sourceFilesCache = createMap<HostFileInfo>();                 // Cache that stores the source file and version info
         let missingFilePathsRequestedForRelease: Path[];                    // These paths are held temparirly so that we can remove the entry from source file cache if the file is not tracked by missing files
         let hasChangedCompilerOptions = false;                              // True if the compiler options have changed between compilations
         let hasChangedAutomaticTypeDirectiveNames = false;                  // True if the automatic type directives have changed
@@ -627,11 +636,20 @@ namespace ts {
             return ts.toPath(fileName, currentDirectory, getCanonicalFileName);
         }
 
+        function isFileMissingOnHost(hostSourceFile: HostFileInfo): hostSourceFile is FileMissingOnHost {
+            return isString(hostSourceFile);
+        }
+
+        function isFilePresentOnHost(hostSourceFile: FileMayBePresentOnHost): hostSourceFile is FilePresentOnHost {
+            return !!(hostSourceFile as FilePresentOnHost).sourceFile;
+        }
+
         function fileExists(fileName: string) {
             const path = toPath(fileName);
-            const hostSourceFileInfo = sourceFilesCache.get(path);
-            if (hostSourceFileInfo !== undefined) {
-                return !isString(hostSourceFileInfo);
+            // If file is missing on host from cache, we can definitely say file doesnt exist
+            // otherwise we need to ensure from the disk
+            if (isFileMissingOnHost(sourceFilesCache.get(path))) {
+                return true;
             }
 
             return directoryStructureHost.fileExists(fileName);
@@ -640,39 +658,42 @@ namespace ts {
         function getVersionedSourceFileByPath(fileName: string, path: Path, languageVersion: ScriptTarget, onError?: (message: string) => void, shouldCreateNewSourceFile?: boolean): SourceFile {
             const hostSourceFile = sourceFilesCache.get(path);
             // No source file on the host
-            if (isString(hostSourceFile)) {
+            if (isFileMissingOnHost(hostSourceFile)) {
                 return undefined;
             }
 
             // Create new source file if requested or the versions dont match
-            if (!hostSourceFile || shouldCreateNewSourceFile || hostSourceFile.version.toString() !== hostSourceFile.sourceFile.version) {
+            if (!hostSourceFile || shouldCreateNewSourceFile || !isFilePresentOnHost(hostSourceFile) || hostSourceFile.version.toString() !== hostSourceFile.sourceFile.version) {
                 const sourceFile = getNewSourceFile();
                 if (hostSourceFile) {
                     if (shouldCreateNewSourceFile) {
                         hostSourceFile.version++;
                     }
+
                     if (sourceFile) {
-                        hostSourceFile.sourceFile = sourceFile;
+                        // Set the source file and create file watcher now that file was present on the disk
+                        (hostSourceFile as FilePresentOnHost).sourceFile = sourceFile;
                         sourceFile.version = hostSourceFile.version.toString();
-                        if (!hostSourceFile.fileWatcher) {
-                            hostSourceFile.fileWatcher = watchFilePath(host, fileName, onSourceFileChange, path, writeLog);
+                        if (!(hostSourceFile as FilePresentOnHost).fileWatcher) {
+                            (hostSourceFile as FilePresentOnHost).fileWatcher = watchFilePath(host, fileName, onSourceFileChange, path, writeLog);
                         }
                     }
                     else {
                         // There is no source file on host any more, close the watch, missing file paths will track it
-                        hostSourceFile.fileWatcher.close();
+                        if (isFilePresentOnHost(hostSourceFile)) {
+                            hostSourceFile.fileWatcher.close();
+                        }
                         sourceFilesCache.set(path, hostSourceFile.version.toString());
                     }
                 }
                 else {
-                    let fileWatcher: FileWatcher;
                     if (sourceFile) {
-                        sourceFile.version = "1";
-                        fileWatcher = watchFilePath(host, fileName, onSourceFileChange, path, writeLog);
-                        sourceFilesCache.set(path, { sourceFile, version: 1, fileWatcher });
+                        sourceFile.version = intialVersionString;
+                        const fileWatcher = watchFilePath(host, fileName, onSourceFileChange, path, writeLog);
+                        sourceFilesCache.set(path, { sourceFile, version: intialVersion, fileWatcher });
                     }
                     else {
-                        sourceFilesCache.set(path, "0");
+                        sourceFilesCache.set(path, intialVersionString);
                     }
                 }
                 return sourceFile;
@@ -697,20 +718,22 @@ namespace ts {
             }
         }
 
-        function removeSourceFile(path: Path) {
+        function nextSourceFileVersion(path: Path) {
             const hostSourceFile = sourceFilesCache.get(path);
             if (hostSourceFile !== undefined) {
-                if (!isString(hostSourceFile)) {
-                    hostSourceFile.fileWatcher.close();
-                    resolutionCache.invalidateResolutionOfFile(path);
+                if (isFileMissingOnHost(hostSourceFile)) {
+                    // The next version, lets set it as presence unknown file
+                    sourceFilesCache.set(path, { version: Number(hostSourceFile) + 1 });
                 }
-                sourceFilesCache.delete(path);
+                else {
+                    hostSourceFile.version++;
+                }
             }
         }
 
         function getSourceVersion(path: Path): string {
             const hostSourceFile = sourceFilesCache.get(path);
-            return !hostSourceFile || isString(hostSourceFile) ? undefined : hostSourceFile.version.toString();
+            return !hostSourceFile || isFileMissingOnHost(hostSourceFile) ? undefined : hostSourceFile.version.toString();
         }
 
         function onReleaseOldSourceFile(oldSourceFile: SourceFile, _oldOptions: CompilerOptions) {
@@ -721,10 +744,10 @@ namespace ts {
             // there was version update and new source file was created.
             if (hostSourceFileInfo) {
                 // record the missing file paths so they can be removed later if watchers arent tracking them
-                if (isString(hostSourceFileInfo)) {
+                if (isFileMissingOnHost(hostSourceFileInfo)) {
                     (missingFilePathsRequestedForRelease || (missingFilePathsRequestedForRelease = [])).push(oldSourceFile.path);
                 }
-                else if (hostSourceFileInfo.sourceFile === oldSourceFile) {
+                else if ((hostSourceFileInfo as FilePresentOnHost).sourceFile === oldSourceFile) {
                     sourceFilesCache.delete(oldSourceFile.path);
                     resolutionCache.removeResolutionsOfFile(oldSourceFile.path);
                 }
@@ -808,27 +831,12 @@ namespace ts {
 
         function onSourceFileChange(fileName: string, eventKind: FileWatcherEventKind, path: Path) {
             updateCachedSystemWithFile(fileName, path, eventKind);
-            const hostSourceFile = sourceFilesCache.get(path);
-            if (hostSourceFile) {
-                // Update the cache
-                if (eventKind === FileWatcherEventKind.Deleted) {
-                    resolutionCache.invalidateResolutionOfFile(path);
-                    if (!isString(hostSourceFile)) {
-                        hostSourceFile.fileWatcher.close();
-                        sourceFilesCache.set(path, (++hostSourceFile.version).toString());
-                    }
-                }
-                else {
-                    // Deleted file created
-                    if (isString(hostSourceFile)) {
-                        sourceFilesCache.delete(path);
-                    }
-                    else {
-                        // file changed - just update the version
-                        hostSourceFile.version++;
-                    }
-                }
+
+            // Update the source file cache
+            if (eventKind === FileWatcherEventKind.Deleted && sourceFilesCache.get(path)) {
+                resolutionCache.invalidateResolutionOfFile(path);
             }
+            nextSourceFileVersion(path);
 
             // Update the program
             scheduleProgramUpdate();
@@ -856,7 +864,7 @@ namespace ts {
                 missingFilesMap.delete(missingFilePath);
 
                 // Delete the entry in the source files cache so that new source file is created
-                removeSourceFile(missingFilePath);
+                nextSourceFileVersion(missingFilePath);
 
                 // When a missing file is created, we should update the graph.
                 scheduleProgramUpdate();
@@ -885,17 +893,10 @@ namespace ts {
                     const fileOrDirectoryPath = toPath(fileOrDirectory);
 
                     // Since the file existance changed, update the sourceFiles cache
-                    const result = cachedDirectoryStructureHost && cachedDirectoryStructureHost.addOrDeleteFileOrDirectory(fileOrDirectory, fileOrDirectoryPath);
-
-                    // Instead of deleting the file, mark it as changed instead
-                    // Many times node calls add/remove/file when watching directories recursively
-                    const hostSourceFile = sourceFilesCache.get(fileOrDirectoryPath);
-                    if (hostSourceFile && !isString(hostSourceFile) && (result ? result.fileExists : directoryStructureHost.fileExists(fileOrDirectory))) {
-                        hostSourceFile.version++;
+                    if (cachedDirectoryStructureHost) {
+                        cachedDirectoryStructureHost.addOrDeleteFileOrDirectory(fileOrDirectory, fileOrDirectoryPath);
                     }
-                    else {
-                        removeSourceFile(fileOrDirectoryPath);
-                    }
+                    nextSourceFileVersion(fileOrDirectoryPath);
 
                     // If the the added or created file or directory is not supported file name, ignore the file
                     // But when watched directory is added/removed, we need to reload the file list

--- a/src/compiler/watch.ts
+++ b/src/compiler/watch.ts
@@ -488,13 +488,13 @@ namespace ts {
         const watchFilePath = compilerOptions.extendedDiagnostics ? ts.addFilePathWatcherWithLogging : ts.addFilePathWatcher;
         const watchDirectoryWorker = compilerOptions.extendedDiagnostics ? ts.addDirectoryWatcherWithLogging : ts.addDirectoryWatcher;
 
+        const getCanonicalFileName = createGetCanonicalFileName(useCaseSensitiveFileNames);
+        let newLine = updateNewLine();
+
         writeLog(`Current directory: ${currentDirectory} CaseSensitiveFileNames: ${useCaseSensitiveFileNames}`);
         if (configFileName) {
             watchFile(host, configFileName, scheduleProgramReload, writeLog);
         }
-
-        const getCanonicalFileName = createGetCanonicalFileName(useCaseSensitiveFileNames);
-        let newLine = updateNewLine();
 
         const compilerHost: CompilerHost & ResolutionCacheHost = {
             // Members for CompilerHost
@@ -583,7 +583,9 @@ namespace ts {
 
             // Compile the program
             if (loggingEnabled) {
-                writeLog(`CreatingProgramWith::\n  roots: ${JSON.stringify(rootFileNames)}\n  options: ${JSON.stringify(compilerOptions)}`);
+                writeLog(`CreatingProgramWith::`);
+                writeLog(`  roots: ${JSON.stringify(rootFileNames)}`);
+                writeLog(`  options: ${JSON.stringify(compilerOptions)}`);
             }
 
             const needsUpdateInTypeRootWatch = hasChangedCompilerOptions || !program;

--- a/src/compiler/watch.ts
+++ b/src/compiler/watch.ts
@@ -420,8 +420,7 @@ namespace ts {
         }
     }
 
-    const intialVersion = 1;
-    const intialVersionString = "1";
+    const initialVersion = 1;
 
     /**
      * Creates the watch from the host for root files and compiler options
@@ -437,7 +436,7 @@ namespace ts {
             sourceFile: SourceFile;
             fileWatcher: FileWatcher;
         }
-        type FileMissingOnHost = string;
+        type FileMissingOnHost = number;
         interface FilePresenceUnknownOnHost {
             version: number;
         }
@@ -637,7 +636,7 @@ namespace ts {
         }
 
         function isFileMissingOnHost(hostSourceFile: HostFileInfo): hostSourceFile is FileMissingOnHost {
-            return isString(hostSourceFile);
+            return typeof hostSourceFile === "number";
         }
 
         function isFilePresentOnHost(hostSourceFile: FileMayBePresentOnHost): hostSourceFile is FilePresentOnHost {
@@ -683,17 +682,17 @@ namespace ts {
                         if (isFilePresentOnHost(hostSourceFile)) {
                             hostSourceFile.fileWatcher.close();
                         }
-                        sourceFilesCache.set(path, hostSourceFile.version.toString());
+                        sourceFilesCache.set(path, hostSourceFile.version);
                     }
                 }
                 else {
                     if (sourceFile) {
-                        sourceFile.version = intialVersionString;
+                        sourceFile.version = initialVersion.toString();
                         const fileWatcher = watchFilePath(host, fileName, onSourceFileChange, path, writeLog);
-                        sourceFilesCache.set(path, { sourceFile, version: intialVersion, fileWatcher });
+                        sourceFilesCache.set(path, { sourceFile, version: initialVersion, fileWatcher });
                     }
                     else {
-                        sourceFilesCache.set(path, intialVersionString);
+                        sourceFilesCache.set(path, initialVersion);
                     }
                 }
                 return sourceFile;

--- a/src/harness/unittests/tscWatchMode.ts
+++ b/src/harness/unittests/tscWatchMode.ts
@@ -1719,6 +1719,38 @@ namespace ts.tscWatch {
                 return [files[0]];
             }
         });
+
+        it("file is deleted and created as part of change", () => {
+            const projectLocation = "/home/username/project";
+            const file: FileOrFolder = {
+                path: `${projectLocation}/app/file.ts`,
+                content: "var a = 10;"
+            };
+            const fileJs = `${projectLocation}/app/file.js`;
+            const configFile: FileOrFolder = {
+                path: `${projectLocation}/tsconfig.json`,
+                content: JSON.stringify({
+                    "include": [
+                        "app/**/*.ts"
+                    ]
+                })
+            };
+            const files = [file, configFile, libFile];
+            const host = createWatchedSystem(files, { currentDirectory: projectLocation, useCaseSensitiveFileNames: true });
+            createWatchOfConfigFile("tsconfig.json", host);
+            verifyProgram();
+
+            file.content += "var b = 10;";
+
+            host.reloadFS(files, { invokeFileDeleteCreateAsPartInsteadOfChange: true });
+            host.runQueuedTimeoutCallbacks();
+            verifyProgram();
+
+            function verifyProgram() {
+                assert.isTrue(host.fileExists(fileJs));
+                assert.equal(host.readFile(fileJs), file.content + "\n");
+            }
+        });
     });
 
     describe("tsc-watch module resolution caching", () => {

--- a/src/harness/unittests/tscWatchMode.ts
+++ b/src/harness/unittests/tscWatchMode.ts
@@ -1730,7 +1730,7 @@ namespace ts.tscWatch {
             const configFile: FileOrFolder = {
                 path: `${projectLocation}/tsconfig.json`,
                 content: JSON.stringify({
-                    "include": [
+                    include: [
                         "app/**/*.ts"
                     ]
                 })
@@ -1740,7 +1740,7 @@ namespace ts.tscWatch {
             createWatchOfConfigFile("tsconfig.json", host);
             verifyProgram();
 
-            file.content += "var b = 10;";
+            file.content += "\nvar b = 10;";
 
             host.reloadFS(files, { invokeFileDeleteCreateAsPartInsteadOfChange: true });
             host.runQueuedTimeoutCallbacks();

--- a/src/harness/virtualFileSystemWithWatch.ts
+++ b/src/harness/virtualFileSystemWithWatch.ts
@@ -246,8 +246,12 @@ interface Array<T> {}`
     }
 
     export interface ReloadWatchInvokeOptions {
+        /** Invokes the directory watcher for the parent instead of the file changed */
         invokeDirectoryWatcherInsteadOfFileChanged: boolean;
+        /** When new file is created, do not invoke watches for it */
         ignoreWatchInvokedWithTriggerAsFileCreate: boolean;
+        /** Invoke the file delete, followed by create instead of file changed */
+        invokeFileDeleteCreateAsPartInsteadOfChange: boolean;
     }
 
     export class TestServerHost implements server.ServerHost, FormatDiagnosticsHost, ModuleResolutionHost {
@@ -315,12 +319,18 @@ interface Array<T> {}`
                         if (isString(fileOrDirectory.content)) {
                             // Update file
                             if (currentEntry.content !== fileOrDirectory.content) {
-                                currentEntry.content = fileOrDirectory.content;
-                                if (options && options.invokeDirectoryWatcherInsteadOfFileChanged) {
-                                    this.invokeDirectoryWatcher(getDirectoryPath(currentEntry.fullPath), currentEntry.fullPath);
+                                if (options && options.invokeFileDeleteCreateAsPartInsteadOfChange) {
+                                    this.removeFileOrFolder(currentEntry, returnFalse);
+                                    this.ensureFileOrFolder(fileOrDirectory);
                                 }
                                 else {
-                                    this.invokeFileWatcher(currentEntry.fullPath, FileWatcherEventKind.Changed);
+                                    currentEntry.content = fileOrDirectory.content;
+                                    if (options && options.invokeDirectoryWatcherInsteadOfFileChanged) {
+                                        this.invokeDirectoryWatcher(getDirectoryPath(currentEntry.fullPath), currentEntry.fullPath);
+                                    }
+                                    else {
+                                        this.invokeFileWatcher(currentEntry.fullPath, FileWatcherEventKind.Changed);
+                                    }
                                 }
                             }
                         }

